Fix: mWeb/Web Safari - Selecting attachment

kidroca opened the pull request:
@Julesssss

### Details
Safari does not show the attachment picker if the button used to trigger file input disappears as soon as it's clicked
Added small updates to the `CreateMenu` components that preserve existing functionality but also allow to
trigger an action before the CreateMenu popover closes

### Fixed Issues
Fixes #2109 and also fixes: #1159

### Tests
This bug affected mobile Safari, mobile Chrome and Desktop Safari

1. From the home screen
2. Select the plus (+) button
3. Select "Upload Photo"
4. The file picker should appear
  - check the screenshots
  - on mobile I had no problem no a physical devices, but on the Simulator, the keyboard would try to appear and the file picking options would hide under. Pressing done would review them, but this is an issue with autofocus handling. 
5. You can try to select the same file twice to see that #1159 is fixed

### QA Steps
Same as above

Note: there are changes to the Create Menu component - the component used to show the "New Chat", "Group Chat" options on the home screen FAB, and the Attachment Picker component which is also used on "Edit Photo" for the profile picture. 
I've verified they still work, but you should probably test it as well

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
![image](https://user-images.githubusercontent.com/12156624/113610728-2eb68700-9656-11eb-9fdf-2458557dcb05.png)

![image](https://user-images.githubusercontent.com/12156624/113610760-383fef00-9656-11eb-8283-803874aa55b8.png)

#### Mobile Web (iPhone 8 Simulator Safari)
![image](https://user-images.githubusercontent.com/12156624/113610838-53126380-9656-11eb-82b5-e2d220054b28.png)

#### Mobile Web (iPhone 11 Pro Safari) 
![image](https://user-images.githubusercontent.com/12156624/113611277-f95e6900-9656-11eb-8491-db0f55202978.png)

#### Mobile Web (Android Pixel 2 Chrome)
![image](https://user-images.githubusercontent.com/12156624/113611215-e5b30280-9656-11eb-988e-4bb1eff68c2a.png)

kidroca committed:
refactor: Simplify and enhance CreateMenu

Sometimes the selected item need to be used to trigger an action before
the popup closes.
Calling `props.onItemSelected` with the pressed item serves for that
purpose while existing workflows are unaffected

kidroca committed:
fix: ReportActionCompose mWeb attachment picker does not appear

Safari, mobile Safari and mobile Chrome, do not show the attachment picker
because the button that is pressed disappears right away because
it's inside the CreateMenu popup

Keeping the button that will be triggered when `openPicker` is called mounted
By not closing the CreateMenu ahead of time

kidroca committed:
fix: AttachmentPicker web uploading the same file does not work

Related to #1159

kidroca commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -300,6 +299,14 @@ class ReportActionCompose extends React.Component {
     }
 }
 
+const CREATE_MENU_ITEMS = {
+    ATTACHMENT: {
+        icon: Paperclip,
+        text: 'Upload Photo',
+        onSelected: () => {},
+    },
+};
```
I don't know where to put this, it doesn't feel right to have it at the top. Maybe in a separate file...

Julesssss replied:
Good point, but it's probably fine to match our existing usages for now.

kidroca commented:
It might be better to wait for  #2114 to be merged as it also made changes to the Attachment Picker

tgolen commented:
OK, looks like that other PR would be good to merge first. Would you mind adding `[HOLD PR#2114]` to the beginning of the title of this PR? Then when the other PR has been merged, you can remove the HOLD and add a comment here to let me know it's ready for review again.

Julesssss commented:
Other PR is now merged, removing the hold.

Julesssss commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -217,20 +218,18 @@ class ReportActionCompose extends React.Component {
+                                                                onPicked: file => displayFileInModal({file}),
+                                                            });
+
+                                                            this.setMenuVisibility(false);
```
Wouldn't we want to hide LHN for every menu item, rather than just the ATTACHMENT?

kidroca replied:
Yep, will update
`this.setMenuVisibility(false);` - this is the Create menu modal though not LHN

Julesssss commented on src/components/CreateMenu.js:
```diff
@@ -21,7 +21,7 @@ const propTypes = {
             icon: PropTypes.func.isRequired,
             text: PropTypes.string.isRequired,
-            onSelected: PropTypes.func.isRequired,
+            onSelected: PropTypes.func,
```
It's not clear to me why this change is necessary, would you mind explaining?

kidroca replied:
Initially I thought I could use something like this: 

```jsx
menuItems={[
    {
        icon: Paperclip,
        text: 'Upload Photo',
    },
]}
```

I'll revert it back to required though

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/safari-attachment-fix

Conflicts:
	src/pages/home/report/ReportActionCompose.js

kidroca committed:
refactor: ReportActionCompose hide the menu when any item is selected

kidroca committed:
refactor: ReportActionCompose restore menuItems original usage

kidroca committed:
refactor: CreateMenu revert `onSelected` to required prop

kidroca commented on src/components/AttachmentPicker/index.js:
```diff
@@ -35,11 +35,18 @@ class AttachmentPicker extends React.Component {
                 <input
                     hidden
                     type="file"
+                    accept="audio/*;video/*;image/*;application/pdf"
```
If we fine tune it - it can filter/suggest files based on this attribute: 
![image](https://user-images.githubusercontent.com/12156624/113751860-093c8280-9715-11eb-8ba0-cd36d981074f.png)

I don't think it does anything at the moment, with or without it there are no restrictions but I'll just remove it to be safe

kidroca committed:
refactor: AttachmentPicker remove the accept attribute - unneeded

kidroca commented:
Updated
@tgolen @Julesssss synced with master and resolved conflicts, addressed requested changes, this is ready for review

kidroca commented:
Hey sorry I've just noticed but I need to also update the usage of the Attachment Picker on the ProfilePage:
https://github.com/Expensify/Expensify.cash/blob/5ab2683945ae684708bd3505a2810461a047e21e/src/pages/settings/ProfilePage/index.js#L196

```jsx
        const menuItems = [
            {
                icon: Upload,
                text: 'Upload Photo',
                onSelected: () => {
                    setTimeout(() => {
                        openPicker({
                            onPicked: setAvatar,
                        });
                    }, 10);
                },
            },
        ];
```

I should remove the timeout and update the handling there as well

kidroca committed:
refactor: extracted BaseCreateMenu component

Native vs web/desktop implementation of the CreateMenu component

On Safari and mobile browsers the attachment picker would not be triggered
because `input.click()` is invoked after the CreateMenu is closed and
`input` is consider destroyed this is why on Web we need to trigger
the click before the CreateMenu closes

On mobile native on the other hand triggering another modal needs to happen after
CreateMenu closes otherwise the second modal wont appear

kidroca committed:
refactor: ReportActionCompose restore original usage

kidroca committed:
refactor: ProfilePage - update attachment picker usage

kidroca committed:
docs: CreateMenu remove "when the modal closes" to avoid confusion

kidroca committed:
refactor ReportActionCompose restore original usage and remove eslint comment

kidroca commented:
Updated

When I originally added this logic, I tested on Android and I missed the fact that calling 
`openPicker` before the first modal closes would not work on iOS native. on iOS the first
modal needs to close and only then you can call `openPicker`

But the tricky part is on mobile web and Safari invoking an action after the first modal
closes is ignored by the browser because the button that triggers `input.click()` is no longer in the DOM

That's why I've added a configuration for this behavior in CreateMenu 
- on web/mWeb/desktop: menu item `onSelected` actions would be triggered before the modal closes
- on mobile native actions would be triggered after the menu modal closes (same as previous)
- no negative side effects or degradations noticed - tried all the places CreateMenu is used (settings/profile, FAB, chat compose)

tgolen commented on src/components/CreateMenu/BaseCreateMenu.js:
```diff
@@ -26,54 +26,53 @@ const propTypes = {
     ...windowDimensionsPropTypes,
+
+    // Trigger the item `onSelected` action immediately after press or after the modal hides
+    invokeActionImmediately: PropTypes.bool,
```
I think the name of this isn't great... It's not clear to me when you would want the action to be triggered when clicking, or when the modal is closed. What would you think about using two separate props?

- `invokeActionWhenSelected`
- `invokeActionWhenModalClosed`

That would make it very explicit and I think that might clean up a little bit of the logic as well.

kidroca replied:
Will update this tomorrow

kidroca replied:
I've addressed this by adding an `executeActionMode` prop restricted to 2 values - on press or after modal hides
If you provide a wrong value by mistake you'll get a prop type warning (and an error)
IDE Intellisense would pickup and show supported options:
![image](https://user-images.githubusercontent.com/12156624/113847831-86f9a000-97a0-11eb-90f9-4470f9bd8864.png)

I didn't go with two separate props as you could set them both to `true` or both to `false` and it shouldn't be possible to configure to invoke `onSelect` twice - when pressed and after close, or never

kidroca replied:
I'm leaving the comment unresolved so you can verify this works for you

tgolen commented on src/pages/settings/ProfilePage/index.js:
```diff
@@ -193,11 +193,9 @@ class ProfilePage extends Component {
                 icon: Upload,
                 text: 'Upload Photo',
                 onSelected: () => {
-                    setTimeout(() => {
```
Thanks for removing this! It was making me cringe 😁

kidroca committed:
refactor CreateMenu clearer action execution configuration

kidroca committed:
refactor CreateMenu update prop name and add documentation

kidroca commented:
Updated

Julesssss commented:
Would you mind reminding me why we need the native iOS menu? Is this still the case for safari mobile web, even though we launch the next Modal before the first closes?

![Screenshot 2021-04-07 at 15 40 13](https://user-images.githubusercontent.com/10736861/113885233-97694500-97b7-11eb-81d4-8b4e547da63e.png)

kidroca commented:
> Would you mind reminding me why we need the native iOS menu here instead of our custom CreateMenu?

You are seeing a screenshot of the second modal - the popup that appears after you select "Upload Photo" from the options in the CreateMenu. It's appearance is handled by the browser.
I didn't take a screenshot of the Create Menu as it only has 1 option at the moment
My screenshot captures the Menu that prior this PR would not appear on mobile web

Maybe I should add a video for clarity. I'll do that when I get the chance

Julesssss commented:
> Maybe I should add a video for clarity. I'll do that when I get the chance

No need, I'm with you. That's my screenshot btw. 

Gotcha, so it's simply because we have more control on mobile (we can show document picker/image picker/camera). Whereas on web we delegate this to the browser, which displays the native context menu?

kidroca commented:
> Gotcha, so it's simply because we have more control on mobile (we can show document picker/image picker/camera). Whereas on web we delegate this to the browser, which displays the native context menu?

Correct

To completely cover the topic, there's an alternative:

You can have a `<input type="file" capture="camera" />` which would directly start the camera. But this means we need to implement another 2nd modal with the options "Camera" and "Browse" -> camera would trigger the camera, while "Browse" would prompt you to select by popping yet another popup (the one in your screenshot). I guess you see how this would not be a great improvement.

https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/capture

kidroca commented:
@Julesssss @tgolen 
Are you waiting for more changes? I don't think there's anything more I need to do here
@tgolen I've left this unresolved so you can verify the solution suits you: https://github.com/Expensify/Expensify.cash/pull/2229#discussion_r608513268

tgolen commented:
Thanks for the bump, I will get it reviewed again now!

tgolen commented on src/components/CreateMenu/BaseCreateMenu.js:
```diff
@@ -0,0 +1,86 @@
+    // Menu items to be rendered on the list
+    menuItems: PropTypes.arrayOf(
+        PropTypes.shape({
+            icon: PropTypes.func.isRequired,
```
Each of these need inline comments too

kidroca replied:
Yeah, no prob, I'm actually not the original author, I just moved CreateMenu.js -> CreateMenu/BaseCreateMenu.js, but I'll do my best to describe these

tgolen commented on src/components/CreateMenu/BaseCreateMenu.js:
```diff
@@ -0,0 +1,86 @@
+
+    /* Configures when menu item actions are triggered: as soon as pressed or after the modal is closed.
+    * On mobile native we need to wait for the modals to close, while on web as soon as clicked */
+    actionExecutionStrategy: PropTypes.oneOf(['ON_PRESS', 'AFTER_MODAL_CLOSE']),
```
I like this idea (much better than two separate props). Can you please make these constants that live in `CONST.js`?

kidroca replied:
Actually taking a new look and I'll need to change this one last time for the best. I'll post a note on the code below

kidroca commented on src/components/CreateMenu/BaseCreateMenu.js:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @param {object} item - an item from this.props.menuItems
+     */
+    selectItem(item) {
+        this.props.onItemSelected(item);
+
+        switch (this.props.actionExecutionStrategy) {
+            case 'ON_PRESS':
+                item.onSelected();
+                this.onModalHide = () => {};
+                break;
+            case 'AFTER_MODAL_CLOSE':
+                this.onModalHide = () => item.onSelected();
+                break;
+            default:
+                throw new Error(`Unexpected "actionExecutionStrategy" value: ${this.props.actionExecutionStrategy}`);
+        }
+    }
```
`case 'ON_PRESS':` 
This is only performed on web, right - so this can be moved to the web version - index.js 

`case 'AFTER_MODAL_CLOSE'` 
This is only performed on native and can also be moved - index.native.js

This the "base" component does not need to know/implement the actual strategies. But the strategies are applied from the platform specific implementation. No confusing props, no confusing constans, the behavior would be clearly described in the platform specific implementation. 

It's as simple as moving one case to the web versions and the other to the native version

kidroca replied:
Leaving this unresolved so you can confirm this works for you

kidroca committed:
refactor: CreateMenu extract platform specific implementations from base

kidroca committed:
refactor: CreateMenu describe prop Types

Fix `icon` prop - it's not just any function
It should be a react component - elementType

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/safari-attachment-fix

kidroca commented:
Updated

tgolen commented on src/components/CreateMenu/BaseCreateMenu.js:
```diff
@@ -0,0 +1,48 @@
+
+const propTypes = {
+    // Callback fired when the menu is completely closed
+    onModalHide: PropTypes.func,
```
I think this prop has a name that is incorrect according to the comment and the component name. I think `onHide` or `onMenuHide` would be more appropriate.

kidroca replied:
I'll update it. It's not something that is exposed to the consumers `CreateMenu` though

tgolen commented on src/components/MenuItem.js:
```diff
@@ -17,7 +17,7 @@ const propTypes = {
 
     // Icon to display on the left side of component
-    icon: PropTypes.func.isRequired,
+    icon: PropTypes.elementType.isRequired,
```
I've never seen `elementType` used before, and I'm a little unclear on how it's supposed to be used. The only hint I can find in the React docs is:

```
// A React element type (ie. MyComponent).
```

Which doesn't make a whole lot of sense... (I would expect to see "MyComponent" in their example somewhere). Ref: https://reactjs.org/docs/typechecking-with-proptypes.html#proptypes
but for this particular instance, I think `PropTypes.node` might be a better propType?

kidroca replied:
`PropType.elementType` stands for either a stateless func component or a class component constructor e.g. not just any function but a specifically a react component.

It does sound a bit confusing but it exactly represents the requirement.

PropType.node would raise a warning as it stands for an evaluated/rendered content e.g. it's used when you have a prop like this:
```jsx
 menuItems={[
   {
     icon: <Paperclip />,
     text: 'Upload Photo',
     onSelect: () => {},
```

While with `PropType.elementType` the signature is this:

```jsx
 menuItems={[
   {
     icon: Paperclip,
     text: 'Upload Photo',
     onSelect: () => {},
```

The latter would allow you to easily pass more props to `Paperclip` if needed e.g. fill, stroke, other svg props

tgolen replied:
OK, neat! Thanks for explaining that subtle difference. Let's keep it `element` in that case!

kidroca replied:
Here I've followed this to the actual usage. It's used like an element:

https://github.com/Expensify/Expensify.cash/blob/d3f1d28478357e2a66b155e660def191d6ed79a8/src/components/Icon/index.js#L31-L38

kidroca committed:
refactor: BaseCreateMenu rename `onModalHide` to `onMenuHide`

kidroca commented:
Updated

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/safari-attachment-fix

kidroca commented:
Updated
Resolved merge conflicts

kidroca commented:
The test stage fails at "Install cocoapods". There are no changes in package or pods in this PR though

Julesssss reviewed (approved):
LGTM. The test was failing due to a missing Airship dependency, which has now been resolved. After re-running the test it succeeded.
